src/utils.jl

```diff
@@ -134,14 +138,83 @@ julia> chunk(1:10, 3)
  5:8
  9:10
 
-julia> chunk(collect(1:10), 3)
-3-element Vector{SubArray{Int64, 1, Vector{Int64}, Tuple{UnitRange{Int64}}, true}}:
- [1, 2, 3, 4]
- [5, 6, 7, 8]
- [9, 10]
+julia> x = reshape(collect(1:20), (5, 4))
+5×4 Matrix{Int64}:
+ 1   6  11  16
+ 2   7  12  17
+ 3   8  13  18
+ 4   9  14  19
+ 5  10  15  20
+
+julia> xs = chunk(x, 2, dims=1)
+2-element Vector{SubArray{Int64, 2, Matrix{Int64}, Tuple{UnitRange{Int64}, Base.Slice{Base.OneTo{Int64}}}, false}}:
+ [1 6 11 16; 2 7 12 17; 3 8 13 18]
+ [4 9 14 19; 5 10 15 20]
+
+julia> xs[1]
+3×4 view(::Matrix{Int64}, 1:3, :) with eltype Int64:
+ 1  6  11  16
+ 2  7  12  17
+ 3  8  13  18
 ```
 """
-chunk(xs, n) = collect(Iterators.partition(xs, ceil(Int, length(xs)/n)))
+chunk(x, n::Int) = collect(Iterators.partition(x, ceil(Int, length(x) / n)))
+
+function chunk(x::AbstractArray, n::Int; dims::Int=ndims(x))
+    idxs = _partition_idxs(x, n, dims) 
+    [selectdim(x, dims, i) for i in idxs]
+end
+
+# Zygote errors if not iterating over collected partitions in [selectdim(x, dims, i) for i in ids] 
+function _partition_idxs(x, n, dims)
+    bs = ceil(Int, size(x, dims) / n)
+    collect(Iterators.partition(axes(x, dims), bs))
+end
+
+@non_differentiable _partition_idxs(x...)
+
+function rrule(::typeof(chunk), x::AbstractArray, n::Int; dims::Int=ndims(x))
+    # this is the implementation of chunk
+    idxs = _partition_idxs(x, n, dims) 
+    y = [selectdim(x, dims, i) for i in idxs]
+    
+    chunk_pullback(dy) = (NoTangent(), ∇chunk(unthunk(dy), x, idxs, Val(dims)))
+    
+    return y, chunk_pullback
+end
+
+# Similar to ∇eachslice  https://github.com/JuliaDiff/ChainRules.jl/blob/8108a77a96af5d4b0c460aac393e44f8943f3c5e/src/rulesets/Base/indexing.jl#L77
+function ∇chunk(dys, x::AbstractArray, idxs, vd::Val{dim}) where {dim}
+    i1 = findfirst(dy -> dy isa AbstractArray, dys)
+    if i1 === nothing  # all slices are Zero!
+        return _zero_fill!(similar(x, float(eltype(x))))
```

Comment:
I also think this should be
```suggestion
    i1 = findfirst(dy -> !(dy isa AbstractZero), dys)
    if i1 === nothing  # all slices are Zero!
        return _zero_fill!(similar(x, float(eltype(x))))
```
to match `if dys[k] isa AbstractZero` below. (The slice gradients could be thunks.)